pennylane/qnode.py

```diff
@@ -123,6 +123,9 @@ class QNode:
         order=1 (int): The order of the finite difference method to use. ``1`` corresponds
             to forward finite differences, ``2`` to centered finite differences.
         shift=pi/2 (float): the size of the shift for two-term parameter-shift gradient computations
+        cache_state=False (bool): for tensorflow and torch interfaces and adjoint differentiation,
+            this indicates whether to save the device state after the forward pass.  Doing so saves a
+            forward execution. Device state automatically reused with autograd and jax interfaces.

```

Comment:
Is there a reason not to have this on when we're doing adjoint differentiation? If we always want it on, then maybe we don't have this keyword argument and just have the state always cached (provided that adjoint diff method is requested)

Reply:
Given state's might occupy a bit of memory, I don't want the memory requirements to balloon.  We may make `True` the default later, but I thought it best to leave it `False` to begin with.

Reply:
:thinking: I was on the fence about adding `cache_state` as a keyword argument, feeling that it might pollute the keyword arguments of the QNode for a fairly specific use case. However, can agree that it might be nice to have the option to turn off/on this feature due to memory limitations. One suggestion would be to rename to something more specific, e.g., `adjoint_cache` since perhaps `cache_state` might give the impression of something more general than it is.

I'd also suggest making `cache_state=True` by default. If it's off by default, I think most users will not notice and not benefit from the feature, so all the hard work in this PR will not be quite so beneficial. Any users experiencing memory issues can always turn off caching if they have a problem.

Reply:
Actually, now I look again, when `cache_state=True`, we cache the state in the Torch and TF interfaces regardless of the `diff_method`? :thinking:

Shouldn't we have it so that caching only occurs if `diff_method="adjoint"`? We can expand to further use cases going forward.

In other words,
Currently: off by default, but when on caching occurs in TF and Torch interfaces regardless of diff method
My suggestion: on by default, and caching occurs in TF and Torch interfaces only when diff method is adjoint

I understand the caution with off-by-default in the former, but I think it's worth following a smart defaults philosophy.